source/request-as-event-emitter.js

```diff
@@ -188,10 +169,48 @@ module.exports = options => {
 			}
 
 			emitter.emit('request', request);
-		});
+
+			const uploadComplete = () => {
+				request.emit('upload-complete');
+			};
+
+			if (is.nodeStream(options.body)) {
+				options.body.once('end', uploadComplete);
+				options.body.pipe(request);
+				options.body = undefined;
+			} else if (options.body) {
+				request.end(options.body, uploadComplete);
+			} else if (input && (options.method === 'POST' || options.method === 'PUT' || options.method === 'PATCH')) {
+				input.once('end', uploadComplete);
+				input.pipe(request);
+			} else {
+				request.end(uploadComplete);
+			}
+		};
+
+		if (options.cache) {
+			const cacheableRequest = new CacheableRequest(fn.request, options.cache); // TODO: we need a manager for CacheableRequest instances
```

Comment:
Am I missing something or would this work?

```js
const mem = require('mem')

// memoized factory
const createCacheableRequest = mem((request, cache) => new CacheableRequest(request, cache))

if (options.cache) {
  const cacheReq = createCacheableRequest(fn.request, options.cache)(options, handleResponse)
  // ... do stuff ...
}
```

Reply:
Seems good to me. I'll try that. Thank you ;)

Reply:
I tried it, and unfortunately it doesn't work...

```js
z = require('mem')((x, y) => Math.random())
[Function]
> a = new Map()
Map {}
> b = new Map()
Map {}
> z(1, a)
0.7761714954957284
> z(1, b)
0.7761714954957284
```

Although `a` and `b` aren't the same objects, it gives the same results. It causes the tests fail.

Reply:
@sindresorhus is it a bug? ^^^^^

Reply:
Yeah, I just discovered that it JSON-stringifies the args. Darn. I suppose you could put a symbol on the cache to uniquely identify it.

Reply:
Yeah, did already so! :)